README.md
# NAME

App::pscan - a handful network scan swiss tool

# SYNOPSIS

    $ pscan [scantype] [iprange]:[port range]

# DESCRIPTION

App::pscan is a small POE network scanner, it scan tcp/udp protocol of the specified range

# OPTIONS

        scantype can be:
                - tcp
                - udp
                - discover (doesn't need the port range)

# AUTHOR

mudler <mudler@dark-lab.net>

# COPYRIGHT

Copyright 2014- mudler

# LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# SEE ALSO


Changes
Revision history for App-pscan

{{$NEXT}}
	- various fixes

0.02  2014-03-08 16:47:48 CET
	- Can send payloads to nodes (-p or --payload option)
	- enhanced output
	- fixed udp scan stop


lib/App/pscan/Command.pm
package App::pscan::Command;
use warnings;
use strict;
use App::pscan;
use base qw(App::CLI App::CLI::Command);

use constant global_options => ( 'help' => 'help' );

sub alias { (
        "t"  => "tcp",
        "u" => "udp",
        "d" => "discover"

        
    ) }

sub invoke {
    my ($pkg, $cmd, @args) = @_;
    local *ARGV = [$cmd, @args];
    my $ret = eval {
        $pkg->dispatch();
    };
    if( $@ ) {
        warn $@;
    }
}

sub run(){
    my $self=shift;
        $self->global_help if ($self->{help});
}

sub global_help {
    print <<'END';
App::pscan
____________

help (--help for full)
    - show help message

tcp [ip, iprange]:[port] (payload)
    - try to send the payload

udp [ip, iprange]:[port] (payload)
    - try to send the payload

discover [ip, iprange]
    - try to discover hosts in the network

END
}

1;


lib/App/pscan/Scanner.pm
package App::pscan::Scanner;
use warnings;
use strict;
use Net::IP;
use App::pscan::Utils;

sub run {
    my $self = shift;
    $self->_gen_range(shift);
    $self->scan();

}

sub _gen_range() {
    my $self = shift;
    my $cmd  = shift;
    my ( $Ip, $Port )
        = ( defined($cmd) and $cmd =~ /:/ )
        ? split( /:/, $cmd )
        : ( $cmd, undef );

    if ( my $IP = new Net::IP($Ip) ) {
        $Ip = $IP;
    }
    elsif ( defined($Ip) ) {
        info "<? Resolving ?> $Ip";
        $Ip = resolve($Ip);
        $Ip = new Net::IP($Ip);
    }

    die( error "-## No ip/hostname to scan ##-" ) if ( !defined($Ip) );
    $self->{'IP'} = $Ip;
    info '-> starting scan <-';
    if ($Port) {
        my ( $f, $l ) = generate_ports($Port);
        $self->{'first'} = $f;
        $self->{'last'}  = $l;
        info 'Scanning for ' . ( ( $l + 1 ) - $f ) . ' port(s)';
    }
}

1;


lib/App/pscan/Command/Udp.pm
package App::pscan::Command::Udp;
use warnings;
use strict;
use base qw( App::pscan::Scanner App::pscan::Command);
use IO::Socket::INET;
use POE;
use constant DATAGRAM_MAXLEN => 1024;
use App::pscan::Utils;

=head1 NAME

App::pscan::Command::udp - test the ip with the udp protocol

=head1 DESCRIPTION

udp scan of a given range of the format of Net::IP and a port range.
e.g.: 192.168.1.0/24:80
      192.168.1.1:20-90
      www.google.it:70-80


=head1 OPTIONS

-p or --payload specify a payload to send within the request

=cut

sub options {
    (   "verbose"     => "verbose",
        "p|payload=s" => "payload"
    );
}

sub scan {
    my $self = shift;
    die("No port given") unless (exists $self->{'first'});
    info 'UDP for '
        . $self->{'IP'}->ip()
        . ' port range: '
        . $self->{'first'} . "-"
        . $self->{'last'};

    my $Payload = $self->{'payload'} || "This is a test by App::pscan";
    info 'Payload: ' . $Payload;
    do {
        for ( $self->{'first'} .. $self->{'last'} ) {
            my $port = $_;
            my $host = $self->{'IP'}->ip();
            POE::Session->create(
                inline_states => {
                    _start       => \&client_start,
                    get_datagram => \&client_read,
                    timed_out    => \&timeout
                },
                args => [ $host, $port, $Payload ],
            );

        }
    } while ( ++$self->{'IP'} );

    # Run the clients until the last one has shut down.
    POE::Kernel->run();
    exit;
}

sub client_start {
    my ( $kernel, $session, $heap, $ip, $port, $Payload )
        = @_[ KERNEL, SESSION, HEAP, ARG0, ARG1, ARG2 ];
    my $socket = IO::Socket::INET->new( Proto => 'udp', Timeout => 4 );
    die "Couldn't create client socket: $!" unless $socket;
    $kernel->select_read( $socket, "get_datagram" );
    $kernel->delay( timed_out => 4, $socket );

    #info "Sending '$message' to $ip, waiting for responses";
    my $server_address = pack_sockaddr_in( $port, inet_aton($ip) );
    if ( my $result = send( $socket, $Payload, 0, $server_address ) ) {
        if ( $result == length($Payload) ) {
            info "Message sent to $ip:$port";
        }
        else {
            die "Trouble sending message: $!";
        }
    }
}

sub client_read {
    my ( $kernel, $socket ) = @_[ KERNEL, ARG0 ];
    my $remote_address
        = recv( $socket, my $message = "", DATAGRAM_MAXLEN, 0 );
    return unless defined $remote_address;
    my ( $peer_port, $peer_addr ) = unpack_sockaddr_in($remote_address);
    my $human_addr = inet_ntoa($peer_addr);

    notice "(answer) $human_addr:$peer_port ...";
    notice ",----- Banner from $human_addr:$peer_port";
    notice "| $message";
    notice "`-----";
    $kernel->select_read($socket);
    $kernel->yield("shutdown");

}

sub timeout {

    $_[KERNEL]->select_read( $_[ARG0] );
    $_[KERNEL]->yield("shutdown");

}
1;


lib/App/pscan/Command/Tcp.pm
package App::pscan::Command::Tcp;
use warnings;
use strict;
use base qw( App::pscan::Scanner App::pscan::Command);
use POE qw(Wheel::SocketFactory Wheel::ReadWrite);
use POE::Component::Client::TCP;
use POE::Filter::Stream;
use App::pscan::Utils;

=head1 NAME

App::pscan::Command::tcp - test the ip with the tcp protocol

=head1 DESCRIPTION

tcp scan of a given range of the format of Net::IP and a port range.
e.g.: 192.168.1.0/24:80
      192.168.1.1:20-90
      www.google.it:70-80

=head1 OPTIONS

-p or --payload specify a payload to send within the request

=cut

sub options {
    (   "verbose"     => "verbose",
        "p|payload=s" => "payload"
    );
}


sub scan() {
    my $self = shift;
    die("No port given") unless (exists $self->{'first'});
    info 'TCP for '
        . $self->{'IP'}->ip()
        . ' port range: '
        . $self->{'first'} . "-"
        . $self->{'last'};

    my $Payload = $self->{'payload'} || "This is a test by App::pscan";
    info 'Payload: '.$Payload;
    do {
        for ( $self->{'first'} .. $self->{'last'} ) {
            my $port = $_;
            my $host = $self->{'IP'}->ip() if exists $self->{'IP'};
            POE::Component::Client::TCP->new(
                RemoteAddress => $host,
                RemotePort    => $port,
                Filter        => "POE::Filter::Stream",

              # The client has connected.  Display some status and prepare to
              # gather information.  Start a timer that will send ENTER if the
              # server does not talk to us for a while.
                Connected => sub {
                    info "connected to $host:$port ...";
                    $_[HEAP]->{banner_buffer} = [];
                    $_[KERNEL]->delay( send_enter => 5 );
                },

                # The connection failed.
                ConnectError => sub {

                    #error "could not connect to $host:$port ...";
                },

              # The server has sent us something.  Save the information.  Stop
              # the ENTER timer, and begin (or refresh) an input timer.  The
              # input timer will go off if the server becomes idle.
                ServerInput => sub {
                    my ( $kernel, $heap, $input ) = @_[ KERNEL, HEAP, ARG0 ];
                    notice "got input from $host:$port ...";
                    push @{ $heap->{banner_buffer} }, $input;
                    $kernel->delay( send_enter    => undef );
                    $kernel->delay( input_timeout => 1 );
                },

                # These are handlers for additional events not included in the
                # default Server::TCP module.  In this example, they handle
                # timers that have gone off.
                InlineStates =>
                    { # The server has not sent us anything yet.  Send an ENTER
                     # keystroke (really a network newline, \x0D\x0A), and wait
                     # some more.
                    send_enter => sub {
                        info "sending enter on $host:$port ...";
                        $_[HEAP]->{server}->put($Payload)
                            if $_[HEAP]->{server};    # sends enter
                        $_[KERNEL]->delay( input_timeout => 5 );
                    },

                # The server sent us something already, but it has become idle
                # again.  Display what the server sent us so far, and shut
                # down.
                    input_timeout => sub {
                        my ( $kernel, $heap ) = @_[ KERNEL, HEAP ];
                        notice "got input timeout from $host:$port ...";
                        notice ",----- Banner from $host:$port";
                        foreach ( @{ $heap->{banner_buffer} } ) {
                            notice "| $_";

                            # print "| ", unpack("H*", $_), "\n";
                        }
                        notice "`-----";
                        $kernel->yield("shutdown");
                    },
                    },
            );
        }
    } while ( ++$self->{'IP'} );

    info 'Spawning scans';

    # Run the clients until the last one has shut down.
    $poe_kernel->run();
    exit;
}

1;
